java/arcs/core/data/expression/Expression.kt

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.data.expression
+
+/**
+ * A DSL for expressions used by queries, refinements, and adapters. Instances can be constructed
+ * directly, from protos or deserialization, and eventually, from a Kotlin implementation of
+ * the Arcs Manifest Parser. A number of operator overloads exist to make hand construction
+ * terse and ergonomic, see Builders.kt.
+ */
+sealed class Expression<T> {
+
+    /**
+     * Implementors denote sub-properties that may be looked up. This is not necessarily limited
+     * to entities, but could also be used for coercions, e.g. 'numberField.integerPart'.
+     */
+    interface Scope {
+        /** The name of the scope, used for debugging and stringification */
+        val scopeName: String
+
+        /** Lookup an entry in a given scope. */
+        fun <T> lookup(param: String): T
+    }
+
+    /**
+     * Visitor pattern for traversing expressions. Implementations of visitor are used for
+     * evaluation, serialization, and type flow.
+     * @param Result type returned from processing a visitation of an [Expression]
+     */
+    interface Visitor<Result> {
+        /** Called when [UnaryExpression] encountered. */
+        fun <E, T> visit(expr: UnaryExpression<E, T>): Result
+
+        /** Called when [BinaryExpression] encountered. */
+        fun <L, R, T> visit(expr: BinaryExpression<L, R, T>): Result
+
+        /** Called when [FieldExpression] encountered. */
+        fun <E : Scope, T> visit(expr: FieldExpression<E, T>): Result
+
+        /** Called when [QueryParameterExpression] encountered. */
+        fun <T> visit(expr: QueryParameterExpression<T>): Result
+
+        /** Called when [NumberLiteralExpression] encountered. */
+        fun visit(expr: NumberLiteralExpression): Result
```

Comment:
Could the following work to enforce static types? 
We parameterize LiteralExpression, and have separate visitors for each type of literal? 
It seems like the generics needed to support this already exist. It would require that we create separate instances of `Expression` for each literal type. WDYT?

Reply:
... is my understanding of this visitor implementation correct? It seems like different literal types (Number, Boolean, Text, Object, etc.) co-exist within the same type. So, within the same expression, one can interoperate between different literal types.

Reply:
Yes, for future string concentration operation, you could have String + Number = String. 

If you're willing to make each and every possible (T,S)->R a separate visitor method, yes, you could make it 100% static, but once you permit BinaryExpression to return an arbitrary return type, you can't enforce static safety. 

Consider 1 + "Hello".. You have a BinaryExpression that is (Number, String) -> String. But for 1 + 2, you have (Number, Number) -> Number. So visit(BinaryExpression) must have return type String|Number, but Kotlin doesn't support this. 

Maybe I'm wrong, but if you can come up with a type signature that will work, I'd be happy to patch it in as a follow on.

Reply:
Maybe I misunderstood you comment, perhaps your asking why visit(NumberLiteral) can't just return Double? The issue is double-dispatch. accept() returns visit(), so when invoking Expression.accept(), you expect the result of some visit() method.

That is, left.accept(this) op right.accept(this),  the accept() method can return anything a visit(expression) can return, not just a literal value.